### PR TITLE
For scaling deployments in particular, use a direct object get rather than the /scale API

### PR DESCRIPTION
**Pull request by coderanger**
So that it can use the informer cache for better performance.

This is annoying as a special case but is so common and improves performance so much that I think it's worthwhile to include. Another option would be to majorly increase the QPS rate limit on the scaling API client however that would also increase kube-apiserver load while watches/informers are generally much less impactful.

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)
- [x] Tests have been added
- [x] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs
- [x] Changelog has been updated

Refs #1449

### CODE REVIEW

File: pkg/scaling/executor/scale_scaledobjects.go [hunk 1]
```diff
@@ -16,18 +19,38 @@ func (e *scaleExecutor) RequestScale(ctx context.Context, scaledObject *kedav1al
...
-	if err != nil {
-		logger.Error(err, "Error getting information on the current Scale (ie. replias count) on the scaleTarget")
-		return
+	// Get the current replica count. As a special case, Deployments fetch directly from the object so they can use the informer cache
+	// to reduce API calls. Everything else uses the scale subresource.
+	var currentScale *autoscalingv1.Scale
+	var currentReplicas int32
+	targetRef := scaledObject.Spec.ScaleTargetRef
+	// TODO Should this use a more generic runtime.Object approach using scheme.Scheme.New()?
+	if (targetRef.APIVersion == "" || strings.HasPrefix(targetRef.APIVersion, "apps/")) &&
```
Comment (zroubalik) [inline pkg/scaling/executor/scale_scaledobjects.go:28]: This condition could be simpler if we use `scaledObject.status.ScaleTargetGVKR` https://github.com/kedacore/keda/blob/5819a1ffdde9c9f14efe74eee783dd34718de363/api/v1alpha1/scaledobject_types.go#L88

This should have the correct API, Version and Kind resolved, see:
https://github.com/kedacore/keda/blob/5819a1ffdde9c9f14efe74eee783dd34718de363/pkg/util/gvkr.go#L18

**Review by zroubalik — COMMENTED** [review]
Looking good, just a minor nit.

File: pkg/scaling/executor/scale_scaledobjects.go (hunk 1 above)
Comment (zroubalik) [reply pkg/scaling/executor/scale_scaledobjects.go:28]: And we could handle `StatefulSet` directly as well.

Comment (coderanger) [reply pkg/scaling/executor/scale_scaledobjects.go]: Added both, much cleaner!

File: pkg/scaling/executor/scale_scaledobjects.go [hunk 2]
```diff
@@ -16,18 +18,45 @@ func (e *scaleExecutor) RequestScale(ctx context.Context, scaledObject *kedav1al
...
+	// to reduce API calls. Everything else uses the scale subresource.
+	var currentScale *autoscalingv1.Scale
+	var currentReplicas int32
+	targetGVKR := scaledObject.Status.ScaleTargetGVKR
+	switch {
+	case targetGVKR.Group == "apps" && targetGVKR.Kind == "Deployment":
+		deployment := &appsv1.Deployment{}
+		err := e.client.Get(ctx, client.ObjectKey{Name: targetGVKR.Resource, Namespace: scaledObject.Namespace}, deployment)
+		if err != nil {
+			logger.Error(err, "Error getting information on the current Scale (ie. replias count) on the scaleTarget")
```
Comment (zroubalik) [inline pkg/scaling/executor/scale_scaledobjects.go:31]:
```suggestion
			logger.Error(err, "Error getting information on the current Scale (ie. replicas count) on the scaleTarget")
```

File: pkg/scaling/executor/scale_scaledobjects.go [hunk 3]
```diff
@@ -16,18 +18,45 @@ func (e *scaleExecutor) RequestScale(ctx context.Context, scaledObject *kedav1al
...
+			return
+		}
+		currentReplicas = *deployment.Spec.Replicas
+	case targetGVKR.Group == "apps" && targetGVKR.Kind == "StatefulSet":
+		statefulSet := &appsv1.StatefulSet{}
+		err := e.client.Get(ctx, client.ObjectKey{Name: targetGVKR.Resource, Namespace: scaledObject.Namespace}, statefulSet)
+		if err != nil {
+			logger.Error(err, "Error getting information on the current Scale (ie. replias count) on the scaleTarget")
```
Comment (zroubalik) [inline pkg/scaling/executor/scale_scaledobjects.go:39]:
```suggestion
			logger.Error(err, "Error getting information on the current Scale (ie. replicas count) on the scaleTarget")
```

File: pkg/scaling/executor/scale_scaledobjects.go [hunk 4]
```diff
@@ -16,18 +18,45 @@ func (e *scaleExecutor) RequestScale(ctx context.Context, scaledObject *kedav1al
...
+			return
+		}
+		currentReplicas = *statefulSet.Spec.Replicas
+	default:
+		var err error
+		currentScale, err = e.getScaleTargetScale(ctx, scaledObject)
+		if err != nil {
+			logger.Error(err, "Error getting information on the current Scale (ie. replias count) on the scaleTarget")
```
Comment (zroubalik) [inline pkg/scaling/executor/scale_scaledobjects.go:47]:
```suggestion
			logger.Error(err, "Error getting information on the current Scale (ie. replicas count) on the scaleTarget")
```

File: pkg/scaling/executor/scale_scaledobjects.go [hunk 5]
```diff
@@ -16,18 +18,45 @@ func (e *scaleExecutor) RequestScale(ctx context.Context, scaledObject *kedav1al
 		"scaledObject.Namespace", scaledObject.Namespace,
 		"scaleTarget.Name", scaledObject.Spec.ScaleTargetRef.Name)
 
-	currentScale, err := e.getScaleTargetScale(ctx, scaledObject)
-	if err != nil {
-		logger.Error(err, "Error getting information on the current Scale (ie. replias count) on the scaleTarget")
-		return
+	// Get the current replica count. As a special case, Deployments fetch directly from the object so they can use the informer cache
```
Comment (zroubalik) [inline pkg/scaling/executor/scale_scaledobjects.go:21]:
```suggestion
	// Get the current replica count. As a special case, Deployments and StatefulSets fetch directly from the object so they can use the informer cache
```

File: pkg/scaling/executor/scale_scaledobjects.go [hunk 6]
```diff
@@ -16,18 +18,45 @@ func (e *scaleExecutor) RequestScale(ctx context.Context, scaledObject *kedav1al
...
+	// to reduce API calls. Everything else uses the scale subresource.
+	var currentScale *autoscalingv1.Scale
+	var currentReplicas int32
+	targetGVKR := scaledObject.Status.ScaleTargetGVKR
+	switch {
+	case targetGVKR.Group == "apps" && targetGVKR.Kind == "Deployment":
+		deployment := &appsv1.Deployment{}
+		err := e.client.Get(ctx, client.ObjectKey{Name: targetGVKR.Resource, Namespace: scaledObject.Namespace}, deployment)
```
Comment (zroubalik) [inline pkg/scaling/executor/scale_scaledobjects.go:29]:
```go
targetName := scaledObject.Spec.ScaleTargetRef.Name
```
...
```suggestion
		err := e.client.Get(ctx, client.ObjectKey{Name: targetName, Namespace: scaledObject.Namespace}, deployment)
```

File: pkg/scaling/executor/scale_scaledobjects.go [hunk 7]
```diff
@@ -16,18 +18,45 @@ func (e *scaleExecutor) RequestScale(ctx context.Context, scaledObject *kedav1al
...
+		if err != nil {
+			logger.Error(err, "Error getting information on the current Scale (ie. replias count) on the scaleTarget")
+			return
+		}
+		currentReplicas = *deployment.Spec.Replicas
+	case targetGVKR.Group == "apps" && targetGVKR.Kind == "StatefulSet":
+		statefulSet := &appsv1.StatefulSet{}
+		err := e.client.Get(ctx, client.ObjectKey{Name: targetGVKR.Resource, Namespace: scaledObject.Namespace}, statefulSet)
```
Comment (zroubalik) [inline pkg/scaling/executor/scale_scaledobjects.go:37]:
```go
targetName := scaledObject.Spec.ScaleTargetRef.Name
```
...
```suggestion
		err := e.client.Get(ctx, client.ObjectKey{Name: targetName, Namespace: scaledObject.Namespace}, statefulSet)
```

File: pkg/scaling/executor/scale_scaledobjects.go (hunk 6 above)
Comment (coderanger) [reply pkg/scaling/executor/scale_scaledobjects.go:29]: Seems weird to use different data sources, increased chance of them being out of sync?

Comment (zroubalik) [reply pkg/scaling/executor/scale_scaledobjects.go:29]: ScaledObject and the target workload should be always in the sync and same namespace. But in your original code, I don't understand the intention, how do you want to guess the name from `targetGVKR.Resource`?

Comment (coderanger) [reply pkg/scaling/executor/scale_scaledobjects.go:29]: I completely misunderstood this code, you are correct :)

**Review by zroubalik — APPROVED** [review]

**Pull request merged**